### PR TITLE
Support friend-declarations for access to internal members of upstream deps

### PR DESCRIPTION
**Pull request by cgruber**
Add more robust support for `friend` declarations of a library or test on another compilation unit, causing it to (a) share a module definition, and (b) have access to the `internal` members of that library. This expands the current feature which exposes this to test only, and generalizes it.

It also addresses an issue in kotlinc (https://youtrack.jetbrains.com/issue/KT-34277) whereby `-Xfriends-path` is interpreted using different path separators between JS and JVM. The present code uses `File.pathSeparator` to obtain the platform-specific separator. However, the JVM backend uses a system which separates path elements with a comma, as a list.

There's also some application of ktlint and buildifier on some files involved in this PR. 

Finally, the dagger example is partitioned into smaller pieces, to illustrate and validate the use of `friend` declarations, directly and transitively.

### CODE REVIEW

File: src/test/kotlin/io/bazel/kotlin/KotlinAssertionTestCase.kt [hunk 1]
```diff
@@ -120,7 +120,7 @@ abstract class BasicAssertionTestCase {
         ProcessBuilder().command("bash", "-c", Paths.get(executable).fileName.toString())
             .also { it.directory(executable.resolveDirectory()) }
             .start().let {
-                it.waitFor(5, TimeUnit.SECONDS)
+                it.waitFor(10, TimeUnit.SECONDS)
```
Comment (inez) [inline src/test/kotlin/io/bazel/kotlin/KotlinAssertionTestCase.kt:123]: Why this change?

Comment (inez) [reply src/test/kotlin/io/bazel/kotlin/KotlinAssertionTestCase.kt:123]: I hope "friends" are not making it half as fast ;)

File: src/main/kotlin/io/bazel/kotlin/builder/tasks/jvm/KotlinJvmTaskExecutor.kt [hunk 2]
```diff
@@ -96,15 +102,18 @@ class KotlinJvmTaskExecutor @Inject internal constructor(
...
+            .map { Paths.get(it).toAbsolutePath() }
+            .joinToString(File.pathSeparator)
         args.addAll(
-            "-cp", inputs.joinedClasspath,
+            "-cp", cp,
             "-api-version", info.toolchainInfo.common.apiVersion,
             "-language-version", info.toolchainInfo.common.languageVersion,
             "-jvm-target", info.toolchainInfo.jvm.jvmTarget,
-            "-Xfriend-paths=${friendPaths.joinToString(File.pathSeparator)}"
+            "-Xfriend-paths=${friendPaths.joinToString(X_FRIENDS_PATH_SEPARATOR)}"
```
Comment (inez) [inline src/main/kotlin/io/bazel/kotlin/builder/tasks/jvm/KotlinJvmTaskExecutor.kt:115]: Just double checking: Should "friends" apply to annotation task (and not only compile task?) (this change is in `getCommonArgs` so it will apply to both).

File: kotlin/internal/jvm/compile.bzl [hunk 3]
```diff
@@ -137,6 +137,21 @@ def _partition_srcs(srcs):
...
 
+def _handle_legacy_friends(ctx):
+    # TODO extract and move this into common. Need to make it generic first.
+    friend = getattr(ctx.attr, "friend", None)
+
+    # Reduce to a single friend if more than one is set (old API)
+    friends = getattr(ctx.attr, "friends", [])
+    if len(friends) == 1:
+        if bool(friend):
+            fail("Can only use friend or friends, but not both. Please only set `friend=`")
```
Comment (inez) [inline kotlin/internal/jvm/compile.bzl:148]: I think this should handle situation when you set `friend` and multiple `friends` (not just one).

File: src/test/kotlin/io/bazel/kotlin/KotlinAssertionTestCase.kt (hunk 1 above)
Comment (cgruber) [reply src/test/kotlin/io/bazel/kotlin/KotlinAssertionTestCase.kt:123]: No, I added some time-delay in the dagger example, using coroutines, and that causes the dagger example to run a few seconds long by design. This just pads it.

File: src/main/kotlin/io/bazel/kotlin/builder/tasks/jvm/KotlinJvmTaskExecutor.kt (hunk 2 above)
Comment (cgruber) [reply src/main/kotlin/io/bazel/kotlin/builder/tasks/jvm/KotlinJvmTaskExecutor.kt:115]: It applies to any access to "internal" - since dagger often needs to see internal things (like `internal` provides methods, it should also be present for annotation processing.

File: kotlin/internal/jvm/compile.bzl (hunk 3 above)
Comment (cgruber) [reply kotlin/internal/jvm/compile.bzl:148]: If you look down one block, setting multiple friends is already disallowed, regardless of whether you used `friend=`.  That was already true in the old code, I'm just re-doing it here. but if you use `friends=` correctly (despite it being deprecated), this second check just makes sure you're not doing both.

File: examples/dagger/test/coffee/BasicTest.kt [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+
+import com.google.common.truth.Truth.assertThat
+import heating.ElectricHeater
+import org.junit.Test
+import time.Delayer
+
+class BasicTest {
+    @Test fun FooTest() {
+        val heater = ElectricHeater(object : Delayer {
+            override fun delay() {
```
Comment (rogerhu) [inline examples/dagger/test/coffee/BasicTest.kt:11]: as i much as i like the dagger heater example i think https://github.com/facebook/buck/pull/2197/files#diff-8dbee0235ce31d15e79389735d429850 made it easier to understand..

File: kotlin/internal/jvm/jvm.bzl [hunk 5]
```diff
@@ -161,6 +161,11 @@ _common_attr = utils.add_dicts(
             default = [],
             allow_files = False,
         ),
+        "friend": attr.label(
+            doc = """A single Kotlin dep which allows this code to access internal members of the given dependency.
+             Currently uses the output jar of the module -- i.e., exported deps won't be included.""",
```
Comment (rogerhu) [inline kotlin/internal/jvm/jvm.bzl:166]: This example helped me understand what you were talking about: https://youtrack.jetbrains.com/issue/KT-34251

**Review by rogerhu — APPROVED** [review]

File: examples/dagger/test/coffee/BasicTest.kt (hunk 4 above)
Comment (inez) [reply examples/dagger/test/coffee/BasicTest.kt:11]: Oh yeah, that Buck's example/test seems great - easy to follow (seems like couldn't be any easier).

**Review by inez — APPROVED** [review]

File: kotlin/internal/jvm/jvm.bzl (hunk 5 above)
Comment (cgruber) [reply kotlin/internal/jvm/jvm.bzl:166]: Also: https://github.com/bazelbuild/rules_kotlin/issues/211

File: examples/dagger/test/coffee/BasicTest.kt (hunk 4 above)
Comment (cgruber) [reply examples/dagger/test/coffee/BasicTest.kt:11]: Possibly, however the example isn't just an example, it's also a test. The buck example doesn't actually adequately exercise the use cases.

Comment (cgruber) [reply examples/dagger/test/coffee/BasicTest.kt:11]: Though, I'll rework the names a bit for clarity.
